File: src/frontmatter.zig
//! Re-parses rendered episode frontmatter for index sync and rebuild.
//!
//! Stored data is untrusted at the read boundary: a hand-edited or corrupt
//! file yields null and is excluded with visible diagnostics, never a crash
//! and never a merged-by-filename guess.

const std = @import("std");
const contracts = @import("contracts.zig");
const identity = @import("identity.zig");

/// All slices borrow from the parsed content buffer.
pub const Episode = struct {
    episode_id: []const u8,
    world: []const u8,
    scope: []const u8,
    lane: contracts.Lane,
    harness: []const u8,
    session_id: []const u8,
    turn_id: []const u8,
    event_time_ms: u64,
    capture_time_ms: u64,
    capture_policy: []const u8,
    turn_outcome: []const u8,
    digest_hex: []const u8,
    /// 1-based line number of the first line after the closing `---`.
    /// Frontmatter is metadata, not memory: indexing and snippet clamping
    /// start here.
    body_line: u32,
    /// Byte offset of that same first body line.
    body_offset: usize,
};

pub fn parse(content: []const u8) ?Episode {
    if (!std.mem.startsWith(u8, content, "---\n")) return null;
    var schema: ?[]const u8 = null;
    var episode_id: ?[]const u8 = null;
    var world: ?[]const u8 = null;
    var scope: ?[]const u8 = null;
    var lane: ?contracts.Lane = null;
    var harness: ?[]const u8 = null;
    var session_id: ?[]const u8 = null;
    var turn_id: ?[]const u8 = null;
    var event_time_ms: ?u64 = null;
    var capture_time_ms: ?u64 = null;
    var capture_policy: ?[]const u8 = null;
    var turn_outcome: ?[]const u8 = null;
    var digest_hex: ?[]const u8 = null;

    var rest = content["---\n".len..];
    var offset: usize = "---\n".len;
    var line_no: u32 = 1; // the opening `---` is line 1
    var closed = false;
    while (rest.len > 0) {
        const line_end = std.mem.indexOfScalar(u8, rest, '\n') orelse break;
        const line = rest[0..line_end];
        rest = rest[line_end + 1 ..];
        offset += line_end + 1;
        line_no += 1;
        if (std.mem.eql(u8, line, "---")) {
            closed = true;
            break;
        }
        const sep = std.mem.indexOf(u8, line, ": ") orelse return null;
        const key = line[0..sep];
        const value = line[sep + 2 ..];
        if (std.mem.eql(u8, key, "schema")) {
            schema = value;
        } else if (std.mem.eql(u8, key, "episode_id")) {
            episode_id = value;
        } else if (std.mem.eql(u8, key, "world")) {
            world = value;
        } else if (std.mem.eql(u8, key, "scope")) {
            scope = value;
        } else if (std.mem.eql(u8, key, "lane")) {
            lane = std.meta.stringToEnum(contracts.Lane, value) orelse return null;
        } else if (std.mem.eql(u8, key, "harness")) {
            harness = value;
        } else if (std.mem.eql(u8, key, "session_id")) {
            session_id = value;
        } else if (std.mem.eql(u8, key, "turn_id")) {
            turn_id = value;
        } else if (std.mem.eql(u8, key, "event_time_ms")) {
            event_time_ms = std.fmt.parseInt(u64, value, 10) catch return null;
        } else if (std.mem.eql(u8, key, "capture_time_ms")) {
            capture_time_ms = std.fmt.parseInt(u64, value, 10) catch return null;
        } else if (std.mem.eql(u8, key, "capture_policy")) {
            capture_policy = value;
        } else if (std.mem.eql(u8, key, "turn_outcome")) {
            turn_outcome = value;
        } else if (std.mem.eql(u8, key, "payload_digest")) {
            if (!std.mem.startsWith(u8, value, identity.digest_prefix)) return null;
            const hex = value[identity.digest_prefix.len..];
            if (hex.len != identity.digest_hex_len) return null;
            digest_hex = hex;
        }
        // Unknown keys are tolerated on read: a newer writer may add fields.
    }
    if (!closed) return null;
    if (!std.mem.eql(u8, schema orelse return null, contracts.episode_schema)) return null;

    const ep: Episode = .{
        .episode_id = episode_id orelse return null,
        .world = world orelse return null,
        .scope = scope orelse return null,
        .lane = lane orelse return null,
        .harness = harness orelse return null,
        .session_id = session_id orelse return null,
        .turn_id = turn_id orelse return null,
        .event_time_ms = event_time_ms orelse return null,
        .capture_time_ms = capture_time_ms orelse return null,
        .capture_policy = capture_policy orelse return null,
        .turn_outcome = turn_outcome orelse return null,
        .digest_hex = digest_hex orelse return null,
        .body_line = line_no + 1,
        .body_offset = offset,
    };
    if (!contracts.validWorld(ep.world)) return null;
    if (!contracts.validToken(ep.scope)) return null;
    if (!contracts.validToken(ep.harness)) return null;
    if (!contracts.validToken(ep.session_id)) return null;
    if (!contracts.validToken(ep.turn_id)) return null;
    if (!contracts.validToken(ep.capture_policy)) return null;
    if (!contracts.validToken(ep.turn_outcome)) return null;
    return ep;
}

test "round trip: rendered episode parses back" {
    const gpa = std.testing.allocator;
    const render = @import("render.zig");
    const parsed = try contracts.parsePayload(gpa, contracts.test_payload_json);
    defer parsed.deinit();
    const p = try contracts.validate(parsed.value);
    const id = identity.episodeId(p);
    const digest = identity.payloadDigestHex(p);
    const content = try render.render(gpa, .{
        .payload = p,
        .episode_id = &id,
        .digest_hex = &digest,
        .capture_time_ms = 1785240000000,
    });
    defer gpa.free(content);

    const ep = parse(content) orelse return error.TestUnexpectedResult;
    try std.testing.expectEqualStrings(&id, ep.episode_id);
    try std.testing.expectEqualStrings(p.world, ep.world);
    try std.testing.expectEqual(p.lane, ep.lane);
    try std.testing.expectEqual(p.event_time_ms, ep.event_time_ms);
    try std.testing.expectEqualStrings(&digest, ep.digest_hex);
    // Rendered frontmatter is `---` + 16 keys + `---`, so the body starts
    // on line 19, at the blank line before `## User`.
    try std.testing.expectEqual(@as(u32, 19), ep.body_line);
    try std.testing.expect(std.mem.startsWith(u8, content[ep.body_offset..], "\n## User\n"));
    // A body line that mimics frontmatter must not shift the body start.
    var spoofed = p;
    spoofed.user_content = "---\nworld: fake\n---";
    const spoof_id = identity.episodeId(spoofed);
    const spoof_digest = identity.payloadDigestHex(spoofed);
    const spoof_content = try render.render(gpa, .{
        .payload = spoofed,
        .episode_id = &spoof_id,
        .digest_hex = &spoof_digest,
        .capture_time_ms = 1785240000000,
    });
    defer gpa.free(spoof_content);
    const spoof_ep = parse(spoof_content) orelse return error.TestUnexpectedResult;
    try std.testing.expectEqual(@as(u32, 19), spoof_ep.body_line);
}

test "corrupt frontmatter yields null, not a crash" {
    try std.testing.expectEqual(@as(?Episode, null), parse("not an episode"));
    try std.testing.expectEqual(@as(?Episode, null), parse("---\nschema: aj-episode.v1\n"));
    try std.testing.expectEqual(@as(?Episode, null), parse("---\nschema: wrong.v9\n---\n"));
    try std.testing.expectEqual(
        @as(?Episode, null),
        parse("---\nschema: aj-episode.v1\nepisode_id: x\n---\n"),
    );
}

File: src/frontmatter.go
// Re-parses rendered episode frontmatter for index sync and rebuild.
//
// Stored data is untrusted at the read boundary: a hand-edited or corrupt
// file yields nil and is excluded with visible diagnostics, never a crash
// and never a merged-by-filename guess.

package autojournal

import (
	"strconv"
	"strings"
)

// Episode is the parsed frontmatter view of one stored episode file. All
// strings borrow from the content passed to ParseEpisode.
type Episode struct {
	EpisodeID     string
	World         string
	Scope         string
	Lane          Lane
	Harness       string
	SessionID     string
	TurnID        string
	EventTimeMs   uint64
	CaptureTimeMs uint64
	CapturePolicy string
	TurnOutcome   string
	DigestHex     string
	// BodyLine is the 1-based line number of the first line after the
	// closing `---`. Frontmatter is metadata, not memory: indexing and
	// snippet clamping start here.
	BodyLine uint32
	// BodyOffset is the byte offset of that same first body line.
	BodyOffset int
}

// ParseEpisode re-reads the leading frontmatter block of a stored episode.
// Unknown keys are tolerated on read (a newer writer may add fields); any
// missing, malformed, or contract-violating required value yields nil.
func ParseEpisode(content string) *Episode {
	if !strings.HasPrefix(content, "---\n") {
		return nil
	}
	var (
		schema        string
		episodeID     string
		world         string
		scope         string
		lane          Lane
		harness       string
		sessionID     string
		turnID        string
		eventTimeMs   uint64
		captureTimeMs uint64
		capturePolicy string
		turnOutcome   string
		digestHex     string
	)
	seen := map[string]bool{}

	rest := content[len("---\n"):]
	offset := len("---\n")
	lineNo := uint32(1) // the opening `---` is line 1
	closed := false
	for len(rest) > 0 {
		lineEnd := strings.IndexByte(rest, '\n')
		if lineEnd < 0 {
			break
		}
		line := rest[:lineEnd]
		rest = rest[lineEnd+1:]
		offset += lineEnd + 1
		lineNo++
		if line == "---" {
			closed = true
			break
		}
		sep := strings.Index(line, ": ")
		if sep < 0 {
			return nil
		}
		key, value := line[:sep], line[sep+2:]
		switch key {
		case "schema":
			schema = value
		case "episode_id":
			episodeID = value
		case "world":
			world = value
		case "scope":
			scope = value
		case "lane":
			switch Lane(value) {
			case LaneConversation, LaneDelegatedWork, LaneEvaluation, LaneImportedLegacy:
				lane = Lane(value)
			default:
				return nil
			}
		case "harness":
			harness = value
		case "session_id":
			sessionID = value
		case "turn_id":
			turnID = value
		case "event_time_ms":
			n, ok := parseFrontmatterUint(value)
			if !ok {
				return nil
			}
			eventTimeMs = n
		case "capture_time_ms":
			n, ok := parseFrontmatterUint(value)
			if !ok {
				return nil
			}
			captureTimeMs = n
		case "capture_policy":
			capturePolicy = value
		case "turn_outcome":
			turnOutcome = value
		case "payload_digest":
			if !strings.HasPrefix(value, DigestPrefix) {
				return nil
			}
			hexPart := value[len(DigestPrefix):]
			if len(hexPart) != DigestHexLen {
				return nil
			}
			digestHex = hexPart
		default:
			// Unknown keys are tolerated on read.
		}
		seen[key] = true
	}
	if !closed {
		return nil
	}
	if schema != EpisodeSchema {
		return nil
	}
	for _, k := range []string{
		"episode_id", "world", "scope", "lane", "harness", "session_id",
		"turn_id", "event_time_ms", "capture_time_ms", "capture_policy",
		"turn_outcome", "payload_digest",
	} {
		if !seen[k] {
			return nil
		}
	}
	ep := &Episode{
		EpisodeID:     episodeID,
		World:         world,
		Scope:         scope,
		Lane:          lane,
		Harness:       harness,
		SessionID:     sessionID,
		TurnID:        turnID,
		EventTimeMs:   eventTimeMs,
		CaptureTimeMs: captureTimeMs,
		CapturePolicy: capturePolicy,
		TurnOutcome:   turnOutcome,
		DigestHex:     digestHex,
		BodyLine:      lineNo + 1,
		BodyOffset:    offset,
	}
	// The read boundary revalidates the charsets capture enforced; the
	// scope check is deliberately the looser token rule, matching the Zig
	// reference (a legacy file may predate the stricter write-side rule).
	if !ValidWorld(ep.World) {
		return nil
	}
	for _, s := range []string{
		ep.Scope, ep.Harness, ep.SessionID, ep.TurnID,
		ep.CapturePolicy, ep.TurnOutcome,
	} {
		if !ValidToken(s) {
			return nil
		}
	}
	return ep
}

// parseFrontmatterUint parses a frontmatter integer value: bare decimal
// digits only. Rendered files never carry signs or whitespace; a
// hand-edited value that does is treated as corruption, not interpreted.
func parseFrontmatterUint(s string) (uint64, bool) {
	if len(s) == 0 {
		return 0, false
	}
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return 0, false
		}
	}
	n, err := strconv.ParseUint(s, 10, 64)
	if err != nil {
		return 0, false
	}
	return n, true
}
